File: capy/src/e2e-coo/js/dev/capylang/test/ObjectOrientedAnnotations.test.js
const test = require('node:test');
const { assert, generatedModule } = require('./_support');

test('reflects object-oriented interface and trait annotations', () => {
    const { AnnotatedContract } = generatedModule('dev/capylang/test/AnnotatedContract.js');
    const { AnnotatedNaming } = generatedModule('dev/capylang/test/AnnotatedNaming.js');
    const contract = AnnotatedContract.type();
    const naming = AnnotatedNaming.type();

    assert.deepEqual(contract.annotations.map(annotation => annotation.name), ['OoAnnotationMarker', 'OoTypeLabel', 'Deprecated']);
    assert.deepEqual(contract.annotations[0].arguments, []);
    assert.equal(contract.annotations[0].pkg.name, 'ObjectOrientedAnnotationDefinitions');
    assert.equal(annotationArgument(contract.annotations[1], 'label').value.value, 'contract');
    assert.equal(annotationArgument(contract.annotations[2], 'message').value.value, 'use AnnotatedContractV2');
    assert.equal(annotationArgument(contract.annotations[2], 'since').value.value, '');
    assert.deepEqual(contract.methods.map(method => method.name), ['render']);
    assert.deepEqual(contract.methods[0].annotations.map(annotation => annotation.name), ['OoMethodName']);

    assert.deepEqual(naming.annotations.map(annotation => annotation.name), ['OoAnnotationMarker', 'OoTypeLabel', 'Deprecated']);
    assert.equal(annotationArgument(naming.annotations[1], 'label').value.value, 'mixin');
    assert.equal(annotationArgument(naming.annotations[2], 'message').value.value, 'use AnnotatedNamingV2');
    assert.equal(annotationArgument(naming.annotations[2], 'since').value.value, '3.0');
    assert.deepEqual(naming.methods.map(method => method.name), ['prefix']);
});

test('reflects object-oriented class field and method annotations', () => {
    const { AnnotatedWidget } = generatedModule('dev/capylang/test/AnnotatedWidget.js');
    const widget = AnnotatedWidget.type();

    assert.deepEqual(widget.annotations.map(annotation => annotation.name), ['OoAnnotationMarker', 'OoTypeLabel', 'Deprecated']);
    assert.equal(annotationArgument(widget.annotations[1], 'label').value.value, 'entity');
    assert.equal(widget.annotations[2].pkg.name, 'Annotations');
    assert.equal(widget.annotations[2].pkg.path, 'capy/meta_prog');
    assert.equal(annotationArgument(widget.annotations[2], 'message').value.value, 'use AnnotatedView');
    assert.equal(annotationArgument(widget.annotations[2], 'since').value.value, '3.0');
    assert.deepEqual(widget.parents.map(parent => parent.name).sort(), ['AnnotatedContract', 'AnnotatedNaming']);
    assert.deepEqual(widget.fields.map(field => field.name), ['name']);
    assert.deepEqual(widget.fields[0].annotations.map(annotation => annotation.name), ['OoAnnotationMarker', 'OoFieldName', 'Deprecated']);
    assert.deepEqual(widget.fields[0].annotations[0].arguments, []);
    assert.equal(annotationArgument(widget.fields[0].annotations[1], 'value').value.value, 'display_name');
    assert.equal(annotationArgument(widget.fields[0].annotations[2], 'message').value.value, 'use label');
    assert.equal(annotationArgument(widget.fields[0].annotations[2], 'since').value.value, '');
    assert.deepEqual(widget.methods.map(method => method.name), ['render']);
    assert.deepEqual(widget.methods[0].annotations.map(annotation => annotation.name), ['OoMethodName', 'Deprecated']);
    assert.equal(annotationArgument(widget.methods[0].annotations[0], 'value').value.value, 'render');
    assert.equal(annotationArgument(widget.methods[0].annotations[1], 'message').value.value, 'use render_label');
    assert.equal(annotationArgument(widget.methods[0].annotations[1], 'since').value.value, '3.1');
});

function annotationArgument(annotation, name) {
    return annotation.arguments.find(argument => argument.name === name);
}

File: capy/src/e2e-coo/py/dev/capylang/test/object_oriented_annotations_test.py
import os
import pathlib
import sys
import unittest


generated_dir = pathlib.Path(os.environ["CAPY_E2E_COO_PY_GENERATED_DIR"])
sys.path.insert(0, str(generated_dir))

from dev.capylang.test.AnnotatedContract import AnnotatedContract
from dev.capylang.test.AnnotatedNaming import AnnotatedNaming
from dev.capylang.test.AnnotatedWidget import AnnotatedWidget


def annotation_argument(annotation, name):
    return next(argument for argument in annotation.arguments if argument.name == name)


class ObjectOrientedAnnotationsPythonE2ETest(unittest.TestCase):
    def test_reflects_object_oriented_interface_and_trait_annotations(self):
        contract = AnnotatedContract.type()
        naming = AnnotatedNaming.type()

        self.assertEqual([annotation.name for annotation in contract.annotations], ["OoAnnotationMarker", "OoTypeLabel", "Deprecated"])
        self.assertEqual(contract.annotations[0].arguments, [])
        self.assertEqual(contract.annotations[0].pkg.name, "ObjectOrientedAnnotationDefinitions")
        self.assertEqual(annotation_argument(contract.annotations[1], "label").value.value, "contract")
        self.assertEqual(annotation_argument(contract.annotations[2], "message").value.value, "use AnnotatedContractV2")
        self.assertEqual(annotation_argument(contract.annotations[2], "since").value.value, "")
        self.assertEqual([method.name for method in contract.methods], ["render"])
        self.assertEqual([annotation.name for annotation in contract.methods[0].annotations], ["OoMethodName"])

        self.assertEqual([annotation.name for annotation in naming.annotations], ["OoAnnotationMarker", "OoTypeLabel", "Deprecated"])
        self.assertEqual(annotation_argument(naming.annotations[1], "label").value.value, "mixin")
        self.assertEqual(annotation_argument(naming.annotations[2], "message").value.value, "use AnnotatedNamingV2")
        self.assertEqual(annotation_argument(naming.annotations[2], "since").value.value, "3.0")
        self.assertEqual([method.name for method in naming.methods], ["prefix"])

    def test_reflects_object_oriented_class_field_and_method_annotations(self):
        widget = AnnotatedWidget.type()

        self.assertEqual([annotation.name for annotation in widget.annotations], ["OoAnnotationMarker", "OoTypeLabel", "Deprecated"])
        self.assertEqual(annotation_argument(widget.annotations[1], "label").value.value, "entity")
        self.assertEqual(widget.annotations[2].pkg.name, "Annotations")
        self.assertEqual(widget.annotations[2].pkg.path, "capy/meta_prog")
        self.assertEqual(annotation_argument(widget.annotations[2], "message").value.value, "use AnnotatedView")
        self.assertEqual(annotation_argument(widget.annotations[2], "since").value.value, "3.0")
        self.assertEqual(sorted(parent.name for parent in widget.parents), ["AnnotatedContract", "AnnotatedNaming"])
        self.assertEqual([field.name for field in widget.fields], ["name"])
        self.assertEqual(
            [annotation.name for annotation in widget.fields[0].annotations],
            ["OoAnnotationMarker", "OoFieldName", "Deprecated"],
        )
        self.assertEqual(widget.fields[0].annotations[0].arguments, [])
        self.assertEqual(annotation_argument(widget.fields[0].annotations[1], "value").value.value, "display_name")
        self.assertEqual(annotation_argument(widget.fields[0].annotations[2], "message").value.value, "use label")
        self.assertEqual(annotation_argument(widget.fields[0].annotations[2], "since").value.value, "")
        self.assertEqual([method.name for method in widget.methods], ["render"])
        self.assertEqual([annotation.name for annotation in widget.methods[0].annotations], ["OoMethodName", "Deprecated"])
        self.assertEqual(annotation_argument(widget.methods[0].annotations[0], "value").value.value, "render")
        self.assertEqual(annotation_argument(widget.methods[0].annotations[1], "message").value.value, "use render_label")
        self.assertEqual(annotation_argument(widget.methods[0].annotations[1], "since").value.value, "3.1")
